contracts/riskon/UniswapV3RiskOnVault.sol

```diff
@@ -0,0 +1,673 @@
+        if (_redeemShares == _totalShares) {
+            harvest();
+            withdraw(baseMintInfo.tokenId, _redeemShares, _totalShares);
+            withdraw(limitMintInfo.tokenId, _redeemShares, _totalShares);
+            delete baseMintInfo;
+            delete limitMintInfo;
+            uint256 borrowTokenBalance = balanceOfToken(borrowToken);
+            if (currentBorrow > borrowTokenBalance) {
+                IERC20Upgradeable(wantToken).safeApprove(RiskOnConstant.UNISWAP_V3_ROUTER, 0);
+                IERC20Upgradeable(wantToken).safeApprove(RiskOnConstant.UNISWAP_V3_ROUTER, balanceOfToken(wantToken));
+                IUniswapV3(RiskOnConstant.UNISWAP_V3_ROUTER).exactOutputSingle(IUniswapV3.ExactOutputSingleParams(wantToken, borrowToken, fee, address(this), block.timestamp, currentBorrow - borrowTokenBalance, type(uint256).max, 0));
+            }
+            __repay(currentBorrow);
+            __removeCollateral(aCollateralToken, IERC20(aCollateralToken).balanceOf(address(this)));
+        } else {
+            uint256 beforeWantTokenBalance = balanceOfToken(wantToken);
+            uint256 beforeBorrowTokenBalance = balanceOfToken(borrowToken);
+            withdraw(baseMintInfo.tokenId, _redeemShares, _totalShares);
+            withdraw(limitMintInfo.tokenId, _redeemShares, _totalShares);
+            //            uint256 redeemVaultWantTokenBalance = beforeWantTokenBalance * _redeemShares / _totalShares;
+            //            uint256 redeemVaultBorrowTokenBalance = beforeBorrowTokenBalance * _redeemShares / _totalShares;
+            //            uint256 redeemWantTokenBalance = afterWantTokenBalance - beforeWantTokenBalance;
+            uint256 redeemBorrowTokenBalance = balanceOfToken(borrowToken) - beforeBorrowTokenBalance;
+            uint256 redeemCurrentBorrow = currentBorrow * _redeemShares / _totalShares;
+            if (redeemCurrentBorrow > redeemBorrowTokenBalance) {
+                IERC20Upgradeable(wantToken).safeApprove(RiskOnConstant.UNISWAP_V3_ROUTER, 0);
+                IERC20Upgradeable(wantToken).safeApprove(RiskOnConstant.UNISWAP_V3_ROUTER, balanceOfToken(wantToken));
+                IUniswapV3(RiskOnConstant.UNISWAP_V3_ROUTER).exactOutputSingle(IUniswapV3.ExactOutputSingleParams(wantToken, borrowToken, fee, address(this), block.timestamp, redeemCurrentBorrow - redeemBorrowTokenBalance, type(uint256).max, 0));
+            }
+            __repay(redeemCurrentBorrow);
+            __removeCollateral(aCollateralToken, IERC20(aCollateralToken).balanceOf(address(this)) * _redeemShares / _totalShares);
+        }
+        uint256 borrowTokenBalance = balanceOfToken(borrowToken);
+        if (borrowTokenBalance > 0) {
+            IERC20Upgradeable(borrowToken).safeApprove(RiskOnConstant.UNISWAP_V3_ROUTER, 0);
+            IERC20Upgradeable(borrowToken).safeApprove(RiskOnConstant.UNISWAP_V3_ROUTER, borrowTokenBalance);
```

Comment:
can put L345 and 346 into function initialize,approve type(uint256).max amount for gas save
```
 IERC20Upgradeable(borrowToken).safeApprove(RiskOnConstant.UNISWAP_V3_ROUTER, type(uint256).max);
```